dig.go

```diff
@@ -320,6 +328,7 @@ func (n *node) Call(c *Container) error {
 		return errWrapf(err, "constructor %v failed", n.ctype)
 	}
 
+	n.called = true
```

Comment:
Consider moving this up to avoid multiple attempts to call an uncallable function.

Reply:
I originally agreed with this and made the change but I realized that then
we'll have to remember the failure if the function failed. IMO we might as
well keep calling the function if it's failing.

Trying to test that did expose a little bug: a failing function leaves the
container in a bad state: it remembers the original value returned by the
constructor even if it failed.

Fixing and adding test.

Reply:
#166